### PR TITLE
Dogfood immutable file collection

### PR DESCRIPTION
**Pull request by wolfs**
This merges the dogfooding of immutable file collection, again.

I included a fix for the bug introduced in incremental compilation which caused to not detect changes for any classes in packages.

We now pass the source file tree to the compilers and only set the classes to compile on `JavaCompileSpec` when we know the actual classes.

### CODE REVIEW

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java [hunk 1]
```diff
@@ -73,6 +73,10 @@ public JavaCompile() {
         CompileOptions compileOptions = getServices().get(ObjectFactory.class).newInstance(CompileOptions.class);
         this.compileOptions = compileOptions;
         CompilerForkUtils.doNotCacheIfForkingViaExecutable(compileOptions, getOutputs());
+
+        // this mimics the behavior of the Ant javac task (and therefore AntJavaCompiler),
+        // which silently excludes files not ending in .java
+        include("**/*.java");
```
Comment (lptr) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: Why do we need this now?

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java [hunk 2]
```diff
@@ -121,6 +125,7 @@ protected void compile(IncrementalTaskInputs inputs) {
             getPath(),
             (IncrementalTaskInputsInternal) inputs,
             source,
+            getSource(),
```
Comment (lptr) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: Can we make this work without passing both `getSource()` and the raw `source`? It seems like we can just use `getSource()` and detect root directories in the (composite) tree, no? You can call `FileCollectionInternal.visitrRootElements()` perhaps?

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java (hunk 1 above)
Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: I removed the filtering from `NormalizingJavaCompiler` - so we only do it here.

Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: By using `include("**/*.java")`, does that also mean we now ignore empty directories in the source root?

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java (hunk 2 above)
Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: Switched to using `visitRootElements` instead.

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy [hunk 3]
```diff
@@ -471,7 +473,7 @@ compileTestJava.options.incremental = true
 
         then:
         outputs.recompiledClasses("A", "B", "C")
-        output.contains("Cannot infer source root(s) for input with type `ArrayList`. Supported types are `File`, `DirectoryTree` and `SourceDirectorySet`.")
+        output.contains("Cannot infer source root(s) for source `file '${textFile.absolutePath}'`. Only things resolving to directory trees are supported.")
```
Comment (lptr) [inline subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: I'm not sure if this message is useful now to the user. Can we explicitly call out type names as we used to, even if just as examples?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/CompilationSourceDirs.java [hunk 4]
```diff
@@ -28,63 +31,72 @@
...
 
     public boolean canInferSourceRoots() {
-        for (Object source : sources) {
-            if (!canInferSourceRoot(source)) {
-                LOG.info("Cannot infer source root(s) for input with type `{}`. Supported types are `File`, `DirectoryTree` and `SourceDirectorySet`. Unsupported input: {}", source.getClass().getSimpleName(), source);
-                return false;
-            }
+        if (sourceRoots == null) {
+            resolveRoots();
         }
```
Comment (lptr) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/CompilationSourceDirs.java]: The check for `null` and setting the value should be in the same method. I'd just let `resolveRoots()` check for `null` and resolve as needed, and then return the resolved value so the caller can work with it.

File: subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/SourceIncrementalJavaCompilationIntegrationTest.groovy]: I am calling out types again.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/CompilationSourceDirs.java (hunk 4 above)
Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/CompilationSourceDirs.java]: Done.

**Pull request merged**